# **Installation and unzipping and importing libraries**

In [1]:
!pip install pandas matplotlib seaborn
!unzip /content/ml-100k.zip
!pip install surprise

Archive:  /content/ml-100k.zip
   creating: ml-100k/
  inflating: ml-100k/allbut.pl       
  inflating: ml-100k/mku.sh          
  inflating: ml-100k/README          
  inflating: ml-100k/u.data          
  inflating: ml-100k/u.genre         
  inflating: ml-100k/u.info          
  inflating: ml-100k/u.item          
  inflating: ml-100k/u.occupation    
  inflating: ml-100k/u.user          
  inflating: ml-100k/u1.base         
  inflating: ml-100k/u1.test         
  inflating: ml-100k/u2.base         
  inflating: ml-100k/u2.test         
  inflating: ml-100k/u3.base         
  inflating: ml-100k/u3.test         
  inflating: ml-100k/u4.base         
  inflating: ml-100k/u4.test         
  inflating: ml-100k/u5.base         
  inflating: ml-100k/u5.test         
  inflating: ml-100k/ua.base         
  inflating: ml-100k/ua.test         
  inflating: ml-100k/ub.base         
  inflating: ml-100k/ub.test         
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 kB 7.9 MB/s eta

In [19]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from surprise import Reader, Dataset, SVD
from surprise import Dataset
from surprise.model_selection import cross_validate, train_test_split , GridSearchCV
from surprise import accuracy
from surprise import SVD
from surprise import dump

# **Train**

In [40]:
dataset_number = 5
path_to_datasets= "/content/ml-100k/"
# Load user data
users_cols = ['user_id', 'age', 'gender', 'occupation', 'zip_code']
users = pd.read_csv(f'{path_to_datasets}u.user', sep='|', names=users_cols, encoding='latin-1')

# Load ratings data
ratings_cols = ['user_id', 'movie_id' ,'rating', 'unix_timestamp']
ratings = pd.read_csv(f'{path_to_datasets}u{dataset_number}.base', sep='\t', names=ratings_cols, encoding='latin-1')

# Load movies data
movies_cols = ['movie_id', 'title', 'release_date', 'video_release_date', 'IMDb_URL'] + ['genre_' + str(i) for i in range(19)]
movies = pd.read_csv(f'{path_to_datasets}u.item', sep='|', names=movies_cols, encoding='latin-1', usecols=range(24))

df = ratings.merge(movies, left_on='movie_id', right_on='movie_id', how='left')

df = df[['user_id', 'title', 'rating']]


reader = Reader(rating_scale=(1,5))
data = Dataset.load_from_df(df, reader)

train, test = train_test_split(data, test_size=0.01)

model = SVD(n_factors=200)
model.fit(train)

# Save the model to a file
model_filename = f"surprise_svd_model{dataset_number}"
dump.dump(model_filename, predictions=None, algo=model)

# Read The for test

In [31]:
dataset_number = 1
path_to_datasets= "/content/ml-100k/"
# Load user data
users_cols = ['user_id', 'age', 'gender', 'occupation', 'zip_code']
users = pd.read_csv(f'{path_to_datasets}u.user', sep='|', names=users_cols, encoding='latin-1')

# Load ratings data
ratings_cols = ['user_id', 'movie_id' ,'rating', 'unix_timestamp']
ratings = pd.read_csv(f'{path_to_datasets}u{dataset_number}.test', sep='\t', names=ratings_cols, encoding='latin-1')

# Load movies data
movies_cols = ['movie_id', 'title', 'release_date', 'video_release_date', 'IMDb_URL'] + ['genre_' + str(i) for i in range(19)]
movies = pd.read_csv(f'{path_to_datasets}u.item', sep='|', names=movies_cols, encoding='latin-1', usecols=range(24))

df = ratings.merge(movies, left_on='movie_id', right_on='movie_id', how='left')

df = df[['user_id', 'title', 'rating']]

# Load the saved model



In [32]:

reader = Reader(rating_scale=(1,5))
data = Dataset.load_from_df(df, reader)

In [33]:
train, test = train_test_split(data, test_size=0.99)

In [35]:
model_filename = 'surprise_svd_model1'

loaded_model = dump.load(model_filename)[1]
predictions = loaded_model.test(test)
accuracy.rmse(predictions)

RMSE: 0.9560


0.9560059948040892

# **evaluation fuction**

In [48]:
def evaluation2(dataset_number):
  path_to_datasets= "/content/ml-100k/"
  # Load user data
  users_cols = ['user_id', 'age', 'gender', 'occupation', 'zip_code']
  users = pd.read_csv(f'{path_to_datasets}u.user', sep='|', names=users_cols, encoding='latin-1')

  # Load ratings data
  ratings_cols = ['user_id', 'movie_id' ,'rating', 'unix_timestamp']
  ratings = pd.read_csv(f'{path_to_datasets}u{dataset_number}.test', sep='\t', names=ratings_cols, encoding='latin-1')

  # Load movies data
  movies_cols = ['movie_id', 'title', 'release_date', 'video_release_date', 'IMDb_URL'] + ['genre_' + str(i) for i in range(19)]
  movies = pd.read_csv(f'{path_to_datasets}u.item', sep='|', names=movies_cols, encoding='latin-1', usecols=range(24))

  df = ratings.merge(movies, left_on='movie_id', right_on='movie_id', how='left')

  df = df[['user_id', 'title', 'rating']]


  reader = Reader(rating_scale=(1,5))
  data = Dataset.load_from_df(df, reader)
  train, test = train_test_split(data, test_size=0.99)
  model_filename = f"surprise_svd_model{dataset_number}"

  loaded_model = dump.load(model_filename)[1]
  predictions = loaded_model.test(test)
  RMSE = accuracy.rmse(predictions)
  print(f"of dataset {dataset_number}")
  return RMSE
x=evaluation2(4)

RMSE: 0.9387
of dataset 4


dataset1 RMSE = 0.9566

dataset2 RMSE = 0.9418

dataset3 RMSE = 0.9364

dataset4 RMSE = 0.9387

dataset5 RMSE = 0.9385
